### ISSUE TITLE
Project parameters on startup

### ISSUE THREAD

**Issue by Michel-0** [Yoooi0/MultiFunPlayer#111]
Hi,

I think I have an _interesting improvement_ that would make it **easier** for all users to actually use MFP more intuitive.

A possibility to load all relevant MFP **parameters** like:
* Player and player settings like it's source
* Script, script source and parameters
* Channel configuration
* Output type and parameters

Just right when **starting** MFP.
Either by using startup call parameters or some specific type of project file.

I guess most of us have some scrips downloaded and may having a hard time setting up MFP for **each** video and script _individually_.
At least I find it kind of annoying to set up all detailed parameters each time I want to use it.

Having such a launcher-file for each script containing all its necessary MFP settings and just double-click it would be way more convenient.

**Maybe this is already possible and I just don't know?**
Does MFP already support some start call parameters?
I've seen you already have a **JSON configuration file**... being able to replace or modify this JSON on startup is basically what I'm asking for.

If it isn't possible yet, I think a simple to implement way of doing it may be:
* Passing a file path to a JSON as startup call parameter when starting MFP
* The referenced JSON must be a valid MFP config file, which then gets immediately applied.
* This referenced JSON might have the extension `.mfp` 

The downside may be that these "project files" need to be full and valid JSON config files.
Maybe hard for you to validate (if you want to) and maybe version-dependent, so possibly broken after updates.
Instead having those project-files contain only a relevant subset of parameters sounds better, but also takes more effort to implement.

Last question would be how to handle those JSON project files as end-user... I've got two ideas:
* Let them simply copy them out of the MFP directory
* Implement a import & export button within the GUI

I think with a batch script replacing the JSON before starting MFP I can already achieve this behavior right now...? Haven't tried yet.

Kind regards

**Comment by Yoooi0** [comment]
From what I've understand you want per video settings?
So for example if you change some settings while playing a specific video, you want to save those changes and then load them each time that video is playing?

I think all this startup/parameter/project stuff is unnecessary complication, those settings could just be loaded on each video change.

**Comment by Michel-0** [comment]
Kind of per-video-settings, yes.
But also - and I think that's the more important thing - automatically select the video.

So passing just the video path on startup and load video specific settings would also be a solution.

**Comment by Yoooi0** [comment]
So I think you want command line parameters? So you would be able to call:
`MultiFunPlayer.exe /Media "D:\video.mp4" /Settings "D:\custom-settings.json"`
This would also work from shortcuts, so no need for a .mfp files.

> But also - and I think that's the more important thing - automatically select the video.

Still a little confused, why double click a .mfp file when you can double click the video file itself and it will be detected by mfp?

> So passing just the video path on startup and load video specific settings would also be a solution.

This is not that simple since mfp needs to be connected to a video player first before it can instruct the video player to load a file.
So the video file from command line would have to wait for a connection.

**Comment by Michel-0** [comment]
> So I think you want command line parameters? So you would be able to call: `MultiFunPlayer.exe /Media "D:\video.mp4" /Settings "D:\custom-settings.json"` This would also work from shortcuts, so no need for a .mfp files.

Sounds good to me.

> Still a little confused, why double click a .mfp file when you can double click the video file itself and it will be detected by mfp?

I am not aware of this feature and this does not seem to be the case as in V1.23.2+54.bc8d418.
If it would work like that, might also be a good solution.

> This is not that simple since mfp needs to be connected to a video player first before it can instruct the video player to load a file. So the video file from command line would have to wait for a connection.

Exactly that's why I came up with the idea of JSON applying configuration settings.
You basically keep the video resource path as video player parameter within the config, this setting might be used.

By just rely on config settings, the actual video resource link would simply be a setting like any other.

**Comment by Yoooi0** [comment]
> I am not aware of this feature and this does not seem to be the case as in V1.23.2+54.bc8d418.
If it would work like that, might also be a good solution.

For example if you use mpv and set it as a default video player then double clicking will change the video file and it will be detected by mfp.

> Exactly that's why I came up with the idea of JSON applying configuration settings.
> You basically keep the video resource path as video player parameter within the config, this setting might be used.
> 
> By just rely on config settings, the actual video resource link would simply be a setting like any other.

Video path is not really a setting, its a parameter, I dont really like having random parameters stored in config files.
This would have to be a different format. So double clicking .mfp files would have to send a list of commands to mfp.

But TBH I fail to see why would you want this.

**Comment by Michel-0** [comment]
> if you use mpv and set it as a default video player then double clicking will change the video file and it will be detected by mfp.

Do I understand you correctly?
If I got MFP running connected to an MPV instance with a video loaded, MFP should detect when I open another video using MPV?
It does not. It opens up a second instance of MPV. The instance of MPV connected to MFP keeps the old video, nothing changes.
What am I doing wrong?

> But TBH I fail to see why would you want this.

The situation described above is basically the most important use case (yet not the only one).
If it _**would**_ work, it might not be necessary.
But it doesn't work for me, so I'm searching for solutions.

**Comment by Yoooi0** [comment]
Oh right, mpv opens each video in new window.
Well for one you would have to create `mpv.conf` file next to `mpv.exe` with this inside: `input-ipc-server=multifunplayer-mpv`, this will make it so each window can be connected to from mfp.

To make it single process only, I found this:
https://apps.microsoft.com/store/detail/mpv-unofficial/9P3JFR0CLLL6?hl=en-us&gl=us
https://github.com/SilverEzhik/mpv-msix
The `mpv.conf` for that seems to be in `%APPDATA%\mpv\mpv-launcher.conf`. So put the above line in this file.

**Comment by Michel-0** [comment]
Hmmm this `single-instance-behavior` only works if I open a video with the win10 use app blablabla dialogue (or default app).

This thing gets installed as "windows app" here: `C:\Program Files\WindowsApps\519Ezhik.mpvUnofficial_0.34.21002.0_x64__xfkyq1j2kc7zp\mpv`
It needs **administrative rights** to be executed directly from that location.
If executed directly from that location, `single-instance-behavior` doesn't work.

Also MFP throws this error when I try to use this MPF executable:
```
MPV failed with exception:

System.ComponentModel.Win32Exception (5): An error occurred trying to start process 'C:\Program Files\WindowsApps\519Ezhik.mpvUnofficial_0.34.21002.0_x64__xfkyq1j2kc7zp\mpv\mpv.exe' with working directory '<SNIP_MFP_DIR>'. 
Access denied
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()
   at System.Diagnostics.Process.Start(ProcessStartInfo startInfo)
   at MultiFunPlayer.MediaSource.ViewModels.MpvMediaSourceViewModel.RunAsync(CancellationToken token)
```
It also thows the exception `Access denied` if I execute MFP as administrator.
So I guess I can't even launch this specific MPV thing by MFP?

Sooo.... not really a solution I guess?

I just realized a video can be updated by drag & drop into a MVP instance connected to MFP, which also seems to update the script.
But this is quite complicated for streaming videos and when when streaming MFP also can't find the script automatically.

**Comment by Yoooi0** [comment]
You dont have to specify the mpv path in mfp, just leave the path empty and add `input-ipc-server=multifunplayer-mpv` to `%APPDATA%\mpv\mpv-launcher.conf`. If you have mpv window open mfp will be able to connect, either manually or you can have auto-connect enabled.

Definitely dont run the `mpv.exe` from that WindowsApps folder, it should register as default video player and auto launch when you open video files.

> I just realized a video can be updated by drag & drop into a MVP instance connected to MFP, which also seems to update the script.

Yes, this is how you use mpv with mfp by default.

> But this is quite complicated for streaming videos and when when streaming MFP also can't find the script automatically.

There are "Media path modifiers" which you can use to translate streams/urls to fake video files which will enable loading scripts if they are found in script libraries.

**Comment by Michel-0** [comment]
> There are "Media path modifiers" which you can use to translate streams/urls to fake video files

Could you please provide some more information about that?
I can hardly find any relevant topics when searching for "media path modifiers"...

**Comment by Yoooi0** [comment]
![MultiFunPlayer_2023-01-06_21-59-21](https://user-images.githubusercontent.com/63662834/211099016-d7376ee0-148f-4399-a324-1bc68b07680f.png)

Did you manage to setup mpv?
I'm probably going to close this as I dont currently see the use case, and I'll make a new issue for per-video settings.

**Comment by Michel-0** [comment]
I did uninstall the MPV win10-version now.
Using the normal MPV again, `input-ipc-server=multifunplayer-mpv` seems to work fine.

I think I did get it working with a Find/Replace path modifier basically converting the video URL to the corresponding filename.
So I guess this is a good enough solution for me.
I don't really see much potential in a "per-video-settings" feature.

For me it was more important to *initially* set everything up as automated as possible.
So this is why I had the idea with this specific end-user-definable config interface.
So everyone could create and save his/her specific setup once and won't have to setup the whole thing on every restart again.

I leave it up to you now what you're doing with this idea.

**Comment by Yoooi0** [comment]
> For me it was more important to initially set everything up as automated as possible.
So this is why I had the idea with this specific end-user-definable config interface.
So everyone could create and save his/her specific setup once and won't have to setup the whole thing on every restart again.

But MFP already does save every possible setting you can configure in the UI, so I dont understand what do you mean by every restart. Only thing it doesnt do is automatically play a video on startup.
MFP will create a `MultiFunPlayer.config.json` file, if it doesnt then you probably dont have permissions to save in MFP directory, in that case running as administrator might help.

I make MFP to be as hands free as possible by design, meaning once you have everything setup, all you do is run it and it will auto connect to video player and devices with the same settings every time.

> I don't really see much potential in a "per-video-settings" feature.

It will allow users to do small tweaks per video and they will be loaded automatically when playing that video.
For example when a script has not enough range you could increase it in MFP only for that one script/video.
Or if a script is too fast then you could limit the speed for one script/video.
